### PR TITLE
plan: refine join cost model.

### PR DESCRIPTION
**Pull request by hanfei1991**
This is the last step of join auto-select. We define the cost for three different types of join:

1. IndexJoin(l, r), we don't care the count of plan r, just a large factor multiple with count(l).
2. MergeJoin(l, r), we think every record take only one piece of cost, which is count(l) + count(r).
3. HashJoin(l, r) that r is small table. Then the cost of looking up once is log2(r), so the cost is count(l) + count(r) + (count(l) + count(r)) * log2(count(r)).

### CODE REVIEW

File: plan/dag_plan_test.go [hunk 1]
```diff
@@ -196,9 +196,8 @@ func (s *testPlanSuite) TestDAGPlanBuilderJoin(c *C) {
 		},
 		{
 			sql:  "select * from t t1 left outer join t t2 on t1.a = t2.a right outer join t t3 on t1.a = t3.a",
-			best: "RightHashJoin{IndexJoin{TableReader(Table(t))->TableReader(Table(t))}(t1.a,t2.a)->TableReader(Table(t))}(t1.a,t3.a)",
+			best: "MergeJoin{MergeJoin{TableReader(Table(t))->TableReader(Table(t))}(t1.a,t2.a)->TableReader(Table(t))}(t1.a,t3.a)",
```
Comment (shenli) [inline plan/dag_plan_test.go]: Why this is changed to merge join?

Comment (hanfei1991) [inline plan/dag_plan_test.go]: Merge Join take less cost

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@winoros PTAL

**Pull request merged**